File: Java/SpectableJavaExamplesTest/src/test/java/spectable/common/MatchTyped.java
package spectable.common;

import java.util.ArrayList;
import java.util.List;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import production.*;
import records.*;
import calculator.*;

public class MatchTyped {
    public String matchedAddress;
    public AddressComponentsTyped addressComponents;

    public MatchTyped(String matchedAddress, AddressComponentsTyped addressComponents) {
        this.matchedAddress = matchedAddress;
        this.addressComponents = addressComponents;
    }

    public MatchTyped(MatchString s) {
        this.matchedAddress = s.matchedAddress;
        this.addressComponents = new AddressComponentsTyped(s.addressComponents);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof MatchTyped)) return false;
        MatchTyped that = (MatchTyped) o;
        return Objects.equals(matchedAddress, that.matchedAddress)
            && Objects.equals(addressComponents, that.addressComponents);
    }

    @Override
    public int hashCode() {
        return Objects.hash(matchedAddress, addressComponents);
    }

    public MatchString toMatchString() {
        return new MatchString(String.valueOf(matchedAddress), addressComponents.toAddressComponentsString());
    }

    public static List<MatchTyped> fromStringList(List<MatchString> list) {
        List<MatchTyped> result = new ArrayList<>();
        for (MatchString s : list) result.add(new MatchTyped(s));
        return result;
    }

    public static List<MatchString> toStringList(List<MatchTyped> list) {
        List<MatchString> result = new ArrayList<>();
        for (MatchTyped t : list) result.add(t.toMatchString());
        return result;
    }

    /** Internal plumbing for {@link Json}; use toJSON() for JSON text. */
    public Map<String, Object> toJsonValue() {
        Map<String, Object> m = new LinkedHashMap<String, Object>();
        m.put("matchedAddress", matchedAddress);
        m.put("addressComponents", addressComponents == null ? null : addressComponents.toJsonValue());
        return m;
    }

    public String toJSON() { return Json.write(toJsonValue()); }

    /** Internal plumbing for {@link Json}; use fromJSON(String) for JSON text. */
    public static MatchTyped fromJsonValue(Map<String, Object> m) {
        if (m == null) return null;
        return new MatchTyped(
            Json.asString(Json.require(m, "matchedAddress"), "matchedAddress"),
            AddressComponentsTyped.fromJsonValue(Json.asObject(Json.require(m, "addressComponents"), "addressComponents")));
    }

    public static MatchTyped fromJSON(String json) {
        return fromJsonValue(Json.parseObject(json));
    }

    public static String toJSONList(List<MatchTyped> list) {
        List<Object> arr = new ArrayList<Object>();
        if (list != null)
            for (MatchTyped item : list) arr.add(item == null ? null : item.toJsonValue());
        return Json.write(arr);
    }

    public static List<MatchTyped> fromJSONList(String json) {
        List<MatchTyped> result = new ArrayList<MatchTyped>();
        for (Object e : Json.parseArray(json))
            result.add(fromJsonValue(Json.asObject(e, "MatchTyped")));
        return result;
    }
}

File: CSharp/SpecStudioExamplesTestsCSharp/common/MatchTyped.cs
namespace SpecStudioExamplesTestsCSharp.common
{
using System;
using System.Buffers;
using System.Collections.Generic;
using System.Globalization;
using System.Text;
using System.Text.Json;
using production;

    public class MatchTyped
    {
        public string matchedAddress;
        public AddressComponentsTyped addressComponents;

        public MatchTyped(string matchedAddress, AddressComponentsTyped addressComponents)
        {
            this.matchedAddress = matchedAddress;
            this.addressComponents = addressComponents;
        }

        public MatchString ToMatchString()
        {
            return new MatchString(
                Json.ToText(this.matchedAddress),
                this.addressComponents.ToAddressComponentsString()
            );
        }

        public static List<MatchString> ToStringList(List<MatchTyped> list)
        {
            var result = new List<MatchString>();
            foreach (var t in list) result.Add(t.ToMatchString());
            return result;
        }

        public static List<MatchTyped> FromStringList(List<MatchString> list)
        {
            var result = new List<MatchTyped>();
            foreach (var s in list) result.Add(s.ToMatchTyped());
            return result;
        }

        public void WriteJson(Utf8JsonWriter w)
        {
            w.WriteStartObject();
            w.WriteString("matchedAddress", this.matchedAddress);
            w.WritePropertyName("addressComponents");
            this.addressComponents.WriteJson(w);
            w.WriteEndObject();
        }

        public string ToJSON()
        {
            var buffer = new ArrayBufferWriter<byte>();
            using (var w = new Utf8JsonWriter(buffer)) { WriteJson(w); }
            return Encoding.UTF8.GetString(buffer.WrittenSpan);
        }

        public static MatchTyped FromJsonElement(JsonElement m)
        {
            return new MatchTyped(
                Json.AsString(Json.Require(m, "matchedAddress"), "matchedAddress"),
                AddressComponentsTyped.FromJsonElement(Json.Require(m, "addressComponents"))
            );
        }

        public static MatchTyped FromJSON(string json)
        {
            return FromJsonElement(Json.Parse(json));
        }

        public static string ToJSONList(List<MatchTyped> list)
        {
            var buffer = new ArrayBufferWriter<byte>();
            using (var w = new Utf8JsonWriter(buffer))
            {
                w.WriteStartArray();
                foreach (var item in list) item.WriteJson(w);
                w.WriteEndArray();
            }
            return Encoding.UTF8.GetString(buffer.WrittenSpan);
        }

        public static List<MatchTyped> FromJSONList(string json)
        {
            var result = new List<MatchTyped>();
            var root = Json.Parse(json);
            Json.RequireArray(root, "MatchTyped");
            foreach (var e in root.EnumerateArray()) result.Add(FromJsonElement(e));
            return result;
        }

        public override string ToString()
        {
            return $"matchedAddress={matchedAddress}, addressComponents={addressComponents}";
        }

        public override bool Equals(object? obj)
        {
            if (obj is not MatchTyped other) return false;
            return object.Equals(this.matchedAddress, other.matchedAddress)
                && object.Equals(this.addressComponents, other.addressComponents);
        }

        public override int GetHashCode()
        {
            var h = new System.HashCode();
            h.Add(this.matchedAddress);
            h.Add(this.addressComponents);
            return h.ToHashCode();
        }
    }
}
